File: version-management/scripts/release.py
#!/usr/bin/env python3
"""
Release Script - Bump version with Semantic Versioning
Usage: python3 scripts/release.py [patch|minor|major] [--note="description"]
"""

import sys
import os
from datetime import datetime
from version_utils import (
    get_current_version, bump_version, write_version,
    append_version_log, PKG_PATH, VERSION_LOG_PATH
)

# package.json이 없으면 기본값 생성
if not os.path.exists(PKG_PATH):
    import json
    default_pkg = {"name": "onlyone-onechat", "version": "1.0.0",
                   "description": "Onlyone OneChat - Version Management"}
    os.makedirs(os.path.dirname(PKG_PATH), exist_ok=True)
    with open(PKG_PATH, 'w') as f:
        json.dump(default_pkg, f, indent=2)
        f.write('\n')
    print("📦 Created default package.json")

args = sys.argv[1:]
bump_type = 'patch'
note = ''

for arg in args:
    if arg in ('major', 'minor', 'patch'):
        bump_type = arg
    elif arg.startswith('--note='):
        note = arg.replace('--note=', '')
    elif arg.startswith('--type='):
        bump_type = arg.replace('--type=', '')

if bump_type not in ('major', 'minor', 'patch'):
    print(f"❌ Invalid bump type: {bump_type}")
    print("   Usage: python3 scripts/release.py [patch|minor|major] [--note=description]")
    sys.exit(1)


def generate_changelog_entry(from_ver, to_ver, btype, note_text):
    date = datetime.now().strftime('%Y-%m-%d')
    labels = {'major': '🔴 MAJOR', 'minor': '🟡 MINOR', 'patch': '🟢 PATCH'}
    lines = [
        f"## [{to_ver}] - {date}",
        "",
        f"**{labels.get(btype, btype)}** - v{from_ver} → v{to_ver}",
        "",
    ]
    if note_text:
        lines.append(f"- {note_text}")
    lines.append("")
    lines.append("")
    return '\n'.join(lines)


def append_changelog(entry):
    changelog_path = os.path.join(os.path.dirname(__file__), '..', 'CHANGELOG.md')
    header = (
        "# Onlyone OneChat Changelog\n\n"
        "All notable changes to this project will be documented here.\n\n"
    )
    if os.path.exists(changelog_path):
        with open(changelog_path, 'r') as f:
            content = f.read()
    else:
        content = header

    # Insert after header
    if content.startswith('# '):
        lines = content.split('\n')
        # Find first ## line
        insert_idx = None
        for i, line in enumerate(lines):
            if line.startswith('## ') and i > 1:
                insert_idx = i
                break
        if insert_idx:
            lines.insert(insert_idx, entry.rstrip('\n'))
        else:
            lines.append(entry.rstrip('\n'))
        content = '\n'.join(lines)

    with open(changelog_path, 'w') as f:
        f.write(content)
    print("📄 CHANGELOG.md updated")


def main():
    current = get_current_version()
    next_ver = bump_version(current, bump_type)

    print("")
    print("╔══════════════════════════════════════╗")
    print("║   Onlyone OneChat - Release       ║")
    print("╠══════════════════════════════════════╣")
    print(f"║   Previous : v{current['raw']:<19}║")
    print(f"║   New      : v{next_ver['raw']:<19}║")
    print(f"║   Type     : {bump_type:<21}║")
    print(f"║   Date     : {datetime.now().strftime('%Y-%m-%d'):<17}║")
    print("╚══════════════════════════════════════╝")
    print("")

    # 1. Update package.json
    write_version(next_ver)

    # 2. Append version log
    append_version_log(current['raw'], next_ver['raw'], bump_type, note)

    # 3. Update changelog
    entry = generate_changelog_entry(current['raw'], next_ver['raw'], bump_type, note)
    append_changelog(entry)

    # 4. Display UI string
    print(f"🎨 UI Version String: Onlyone OneChat v{next_ver['raw']}")
    print("✅ Release complete!")
    print("")
    print("💡 Next steps:")
    print(f"   python3 scripts/inject_version.py  → Update HTML with new version")
    print(f"   git add . && git commit -m \"chore(release): v{next_ver['raw']}\"")
    print(f"   git tag v{next_ver['raw']}")
    print("")


if __name__ == '__main__':
    main()

File: version-management/scripts/release.js
/**
 * Release Script - Bump version with Semantic Versioning
 * Usage: node scripts/release.js [patch|minor|major] [--note="description"]
 */

const versionUtils = require('./version-utils');

const args = process.argv.slice(2);
const bumpType = args[0] || 'patch';
const noteArg = args.find(a => a.startsWith('--note='));
const note = noteArg ? noteArg.replace('--note=', '') : '';

if (!['major', 'minor', 'patch'].includes(bumpType)) {
  console.error(`❌ Invalid bump type: ${bumpType}`);
  console.error('   Usage: node scripts/release.js [patch|minor|major] [--note="description"]');
  process.exit(1);
}

function main() {
  const current = versionUtils.getCurrentVersion();
  const next = versionUtils.bumpVersion(current, bumpType);

  console.log('');
  console.log('╔══════════════════════════════════════╗');
  console.log('║   Onlyone OneChat - Release       ║');
  console.log('╠══════════════════════════════════════╣');
  console.log(`║   Previous : v${current.raw.padEnd(19)}║`);
  console.log(`║   New      : v${next.raw.padEnd(19)}║`);
  console.log(`║   Type     : ${bumpType.padEnd(21)}║`);
  console.log(`║   Date     : ${new Date().toISOString().split('T')[0].padEnd(17)}║`);
  console.log('╚══════════════════════════════════════╝');
  console.log('');

  // 1. Update package.json
  versionUtils.writeVersion(next);

  // 2. Append version log
  versionUtils.appendVersionLog(current.raw, next.raw, bumpType, note);

  // 3. Generate changelog entry
  const changelogEntry = generateChangelogEntry(current.raw, next.raw, bumpType, note);
  appendChangelog(changelogEntry);

  // 4. Display version string for UI
  console.log(`🎨 UI Version String: Onlyone OneChat v${next.raw}`);
  console.log(`✅ Release complete!`);
  console.log('');
  console.log('💡 Next steps:');
  console.log('   node scripts/inject-version.js  → Update HTML with new version');
  console.log('   git add . && git commit -m "chore(release): v${next.raw}"');
  console.log('   git tag v${next.raw}');
  console.log('');
}

function generateChangelogEntry(from, to, type, note) {
  const date = new Date().toISOString().split('T')[0];
  const typeLabels = { major: '🔴 MAJOR', minor: '🟡 MINOR', patch: '🟢 PATCH' };
  let entry = `## [${to}] - ${date}\n\n`;
  entry += `**${typeLabels[type] || type}** - v${from} → v${to}\n\n`;
  if (note) entry += `- ${note}\n`;
  entry += '\n';
  return entry;
}

function appendChangelog(entry) {
  const fs = require('fs');
  const path = require('path');
  const changelogPath = path.join(__dirname, '..', 'CHANGELOG.md');

  let content = '';
  if (fs.existsSync(changelogPath)) {
    content = fs.readFileSync(changelogPath, 'utf8');
  } else {
    content = '# Changelog\n\nAll notable changes to Onlyone OneChat will be documented in this file.\n\n';
  }

  // Insert after the header
  const lines = content.split('\n');
  const headerEnd = lines.findIndex((l, i) => i > 0 && l.startsWith('## '));
  if (headerEnd > 0) {
    lines.splice(headerEnd, 0, entry);
  } else {
    lines.push(entry);
  }
  fs.writeFileSync(changelogPath, lines.join('\n'));
  console.log(`📄 CHANGELOG.md updated`);
}

main();